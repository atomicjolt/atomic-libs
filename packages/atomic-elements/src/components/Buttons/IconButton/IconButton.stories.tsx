import { Meta, StoryObj } from "@storybook/react";
import { IconButton } from ".";
import buttonMeta from "../Button/Button.stories";
import { getCssProps } from "@sb/cssprops";
import { MaterialSymbol } from "@components/Icons/MaterialSymbol";

const meta: Meta<typeof IconButton> = {
  title: "Buttons/IconButton",
  component: IconButton,
  parameters: {
    layout: "centered",
    cssprops: getCssProps("Button"),
  },
  argTypes: {
    ...buttonMeta.argTypes,
  },
};

export default meta;

type Story = StoryObj<typeof IconButton>;

export const Primary: Story = {
  args: {
    children: <MaterialSymbol symbol="more_vert" />,
    isDisabled: false,
    "aria-label": "More options",
    isLoading: false,
    variant: "border",
  },
};

export const IsLoading: Story = {
  args: {
    ...Primary.args,
    isLoading: true,
    loadingLabel: "Loading...",
    loadingComplete: false,
  },
};
