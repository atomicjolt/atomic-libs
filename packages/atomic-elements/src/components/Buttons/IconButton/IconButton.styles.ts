import styled from "styled-components";
import { Button } from '../Button';

export const StyledIconButton = styled(Button)`
  --size-sm-x: 30px;
  --size-sm-y: 30px;
  --size-md-x: 40px;
  --size-md-y: 40px;
  --size-lg-x: 60px;
  --size-lg-y: 60px;

  --btn-height: var(--size-y);
  position: relative;
  height: var(--btn-height);
  width: var(--btn-height);
  padding: 0 0;

  i {
    margin-left: 0;
  }
`;
