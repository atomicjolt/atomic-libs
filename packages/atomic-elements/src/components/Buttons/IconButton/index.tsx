export { IconButton } from "./IconButton.component";
export type { IconButtonProps } from "./IconButton.component";
