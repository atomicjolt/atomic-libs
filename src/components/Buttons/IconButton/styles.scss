@use "../../defines" as *;
@use "../common.scss";

.aj-btn {
  &--icon {
    --btn-text-clr: var(--text-clr-alt);
    --btn-bg-clr: var(--neutral50);
    --btn-hover-text-clr: var(--text-clr);
    --btn-hover-bg-clr: var(--neutral100);
    --btn-border: var(--border);
    width: var(--btn-height);
    padding: var(--btn-padding);
    i {
      margin-left: 0;
    }
    @extend .aj-btn;
  }
}
