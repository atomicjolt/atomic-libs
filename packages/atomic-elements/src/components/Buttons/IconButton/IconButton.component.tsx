import { forwardRef } from "react";
import { CanHaveIcon } from "../../../types";
import { MaterialIcon } from "../../Icons/MaterialIcon";
import { StyledIconButton } from "./IconButton.styles";
import { ButtonProps } from "../Button";
import { useContextProps } from "@hooks/useContextProps";
import { ButtonContext } from "../Button/Button.context";

export interface IconButtonProps extends ButtonProps, CanHaveIcon {}

/** A button designed for displaying a single icon. The button has no text content,
 * so you should provide an aria-label for accessiblity
 *
 * @example
 * <IconButton aria-label="Add" onPress={() => alert("Hello, world!")}>
 *  <MaterialSybmol symbol="add" />
 * </IconButton>
 * */
export const IconButton = forwardRef<HTMLButtonElement, IconButtonProps>(
  function IconButton(props, forwardedRef) {
    [props, forwardedRef] = useContextProps(
      ButtonContext as any,
      props,
      forwardedRef
    );

    if (!props["aria-label"] && !props["aria-labelledby"]) {
      console.warn(
        "IconButton should have an aria-label or aria-labelledby prop for accessibility"
      );
    }

    const {
      icon,
      iconVariant,
      size = "medium",
      variant = "border",
      as,
      children,
      ...rest
    } = props;

    return (
      // @ts-expect-error - forwardedAs isn't being typed correctly
      <StyledIconButton
        {...rest}
        size={size}
        variant={variant}
        ref={forwardedRef}
        forwardedAs={as}
      >
        {icon ? (
          <MaterialIcon icon={icon} variant={iconVariant} size={size} />
        ) : (
          children
        )}
      </StyledIconButton>
    );
  }
);
