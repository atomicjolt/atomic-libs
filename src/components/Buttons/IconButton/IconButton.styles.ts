import styled from "styled-components";
import { BaseStyledButton } from "../Buttons.styles";

export const StyledIconButton = styled(BaseStyledButton)`
  position: relative;
  height: var(--btn-height);
  width: var(--btn-height);
  padding: 0 0;

  i {
    margin-left: 0;
  }
`;
