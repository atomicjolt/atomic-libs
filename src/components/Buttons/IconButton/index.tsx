import React from "react";
import '../../general.scss';
import "./styles.scss";
import Spinner from '../../Loaders/Spinner';

export interface Props {
  /** What to render within the Button */
  icon: string;
  id?: string;
  ariaLabel: string;
  ariaExpanded?: boolean;
  ariaHasPopup?: string;
  ariaControls?: string;
  disabled?: boolean;
  role?: string;
  onClick?: (e: React.MouseEvent<HTMLButtonElement>) => void;
}

/** Icon Button Component */
export default function IconButton({ 
    icon,
    id,
    ariaLabel, 
    ariaExpanded,
    ariaHasPopup,
    ariaControls,
    disabled,
    onClick 
  }: Props) {
    
  return (
    <button
      id={id}
      className="aj-btn--icon"
      type="button"
      onClick={onClick}
      disabled={disabled}
      aria-label={ariaLabel}
      aria-controls={ariaControls}
      aria-haspopup={ariaHasPopup}
      aria-expanded={ariaExpanded}
    >
      <i className="material-icons" aria-hidden>{icon}</i>
    </button>
  );
}
