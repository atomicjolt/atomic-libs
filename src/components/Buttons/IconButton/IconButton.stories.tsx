import { ComponentMeta, ComponentStory } from "@storybook/react";
import IconButton, { Props } from ".";

export default {
  title: "Buttons/IconButton",
  component: IconButton,
} as ComponentMeta<typeof IconButton>;

const Template: ComponentStory<typeof IconButton> = (args: Props) => (
  <IconButton {...args} />
);

export const Default = Template.bind({});
Default.args = {
  icon: "more_vert",
  ariaLabel: "More options"
};
