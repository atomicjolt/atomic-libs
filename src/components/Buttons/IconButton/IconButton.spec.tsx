import React from "react";
import { describe, test, expect } from "vitest";
import { render } from "@testing-library/react";
import IconButton from ".";

test("matches snapshot", () => {
  const result = render(
    <IconButton ariaLabel="More Button" icon="more_vert" />
  );
  expect(result.asFragment()).toMatchSnapshot();
});
