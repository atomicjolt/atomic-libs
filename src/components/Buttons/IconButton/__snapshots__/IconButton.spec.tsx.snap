// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <button
    aria-label="More Button"
    class="sc-imWZod sc-jXbVAB krOeTw cpWLcV aje-btn aje-btn--icon aje-btn--border is-medium"
    type="button"
  >
    <i
      aria-hidden="true"
      class="sc-kAycRU lbPEgw aje-icon material-icons is-medium"
    >
      more_vert
    </i>
  </button>
</DocumentFragment>
`;
